File: go/problem_64.go
package main

func knightTours(n int) int {
	board := make([][]bool, n)
	for i := range board {
		board[i] = make([]bool, n)
	}
	count := 0
	moves := [][2]int{{2, 1}, {2, -1}, {-2, 1}, {-2, -1}, {1, 2}, {1, -2}, {-1, 2}, {-1, -2}}
	var dfs func(r, c, visited int)
	dfs = func(r, c, visited int) {
		if visited == n*n {
			count++
			return
		}
		for _, m := range moves {
			nr, nc := r+m[0], c+m[1]
			if nr >= 0 && nr < n && nc >= 0 && nc < n && !board[nr][nc] {
				board[nr][nc] = true
				dfs(nr, nc, visited+1)
				board[nr][nc] = false
			}
		}
	}
	for r := 0; r < n; r++ {
		for c := 0; c < n; c++ {
			board[r][c] = true
			dfs(r, c, 1)
			board[r][c] = false
		}
	}
	return count
}

func problem64() {
	println("Problem 64:")
	assertEqual(knightTours(1), 1)
	assertEqual(knightTours(2), 0)
	assertEqual(knightTours(3), 0)
	println("All Tests Passed")
}

File: rust/problem_64.rs
pub fn knight_tours(n: usize) -> usize {
    let mut board = vec![vec![false; n]; n];
    let moves: &[(i64,i64)] = &[(2,1),(2,-1),(-2,1),(-2,-1),(1,2),(1,-2),(-1,2),(-1,-2)];
    let mut count = 0;
    fn dfs(r: i64, c: i64, visited: usize, n: usize, board: &mut Vec<Vec<bool>>, moves: &[(i64,i64)], count: &mut usize) {
        if visited == n*n { *count += 1; return; }
        for &(dr, dc) in moves {
            let (nr, nc) = (r+dr, c+dc);
            if nr>=0 && nr<n as i64 && nc>=0 && nc<n as i64 && !board[nr as usize][nc as usize] {
                board[nr as usize][nc as usize] = true;
                dfs(nr, nc, visited+1, n, board, moves, count);
                board[nr as usize][nc as usize] = false;
            }
        }
    }
    for r in 0..n { for c in 0..n {
        board[r][c] = true;
        dfs(r as i64, c as i64, 1, n, &mut board, moves, &mut count);
        board[r][c] = false;
    } }
    count
}

pub fn problem64() {
    println!("Problem 64:");
    assert_eq!(knight_tours(1), 1);
    assert_eq!(knight_tours(2), 0);
    assert_eq!(knight_tours(3), 0);
    println!("All Tests Passed");
}
